rust/headless-client/src/dns_control/windows.rs

```diff
@@ -82,49 +78,50 @@ pub(crate) fn system_resolvers() -> Result<Vec<IpAddr>> {
     Ok(resolvers)
 }
 
+// Our NRPT rule should always live at
+// `Computer\HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\Dnscache\Parameters\DnsPolicyConfig\{6C0507CB-C884-4A78-BC55-0ACEE21227F6}`
+const NRPT_REG_KEY: &str = "{6C0507CB-C884-4A78-BC55-0ACEE21227F6}";
+
 /// Tells Windows to send all DNS queries to our sentinels
 ///
 /// Parameters:
 /// - `dns_config_string`: Comma-separated IP addresses of DNS servers, e.g. "1.1.1.1,8.8.8.8"
-pub(crate) fn activate(dns_config: &[IpAddr]) -> Result<()> {
-    let dns_config_string = dns_config
-        .iter()
-        .map(|ip| format!("\"{ip}\""))
-        .collect::<Vec<_>>()
-        .join(",");
-
-    // Set our DNS IP as the DNS server for our interface
+fn activate(dns_config: &[IpAddr]) -> Result<()> {
     // TODO: Known issue where web browsers will keep a connection open to a site,
     // using QUIC, HTTP/2, or even HTTP/1.1, and so they won't resolve the DNS
     // again unless you let that connection time out:
     // <https://github.com/firezone/firezone/issues/3113#issuecomment-1882096111>
-    Command::new("powershell")
-        .creation_flags(CREATE_NO_WINDOW)
-        .arg("-Command")
-        .arg(format!(
-            "Set-DnsClientServerAddress {TUNNEL_NAME} -ServerAddresses({dns_config_string})"
-        ))
-        .status()?;
 
     tracing::info!("Activating DNS control");
-    Command::new("powershell")
-        .creation_flags(CREATE_NO_WINDOW)
-        .args([
-            "-Command",
-            "Add-DnsClientNrptRule",
-            "-Namespace",
-            ".",
-            "-Comment",
-            FZ_MAGIC,
-            "-NameServers",
-            &dns_config_string,
-        ])
-        .status()?;
+    let dns_config_string = dns_config
+        .iter()
+        .map(|ip| ip.to_string())
+        .collect::<Vec<_>>()
+        .join(";");
+
+    let hkcu = winreg::RegKey::predef(winreg::enums::HKEY_LOCAL_MACHINE);
+    let base = Path::new("SYSTEM")
```

Comment:
A candidate for the upcoming `LazyLock`: https://github.com/rust-lang/rust/pull/121377